staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope_test.go

```diff
@@ -890,6 +891,112 @@ func TestEnvelopeMetrics(t *testing.T) {
+	envelopeService := newTestEnvelopeService()
+	// to avoid data race
+	var testState atomic.Pointer[State]
+	state1, err := testStateFunc(testContext(t), envelopeService, clock.RealClock{}, true)()
+	if err != nil {
+		t.Fatal(err)
+	}
+	testState.Store(&state1)
+	// creating more states to ensure we get more than one cache key for a given transformer
+	state2, err := testStateFunc(testContext(t), envelopeService, clock.RealClock{}, false)()
+	if err != nil {
+		t.Fatal(err)
+	}
+	state3, err := testStateFunc(testContext(t), envelopeService, clock.RealClock{}, false)()
+	if err != nil {
+		t.Fatal(err)
+	}
+	ctx := testContext(t)
+	dataCtx := value.DefaultContext(testContextText)
+	provider1 := "one"
+	provider2 := "two"
+
+	testCases := []struct {
+		desc                  string
+		keyVersionFromEncrypt string
+		metrics               []string
+		want                  string
+	}{
+		{
+			desc:                  "dek source cache size",
+			keyVersionFromEncrypt: testKeyVersion,
+			metrics: []string{
+				"apiserver_envelope_encryption_dek_source_cache_size",
+			},
+			want: fmt.Sprintf(`
+				# HELP apiserver_envelope_encryption_dek_source_cache_size [ALPHA] Number of records in data enecryption key (DEK) source cache.
+				# TYPE apiserver_envelope_encryption_dek_source_cache_size gauge
+        		apiserver_envelope_encryption_dek_source_cache_size{provider_name="%s"} 3
+        		apiserver_envelope_encryption_dek_source_cache_size{provider_name="%s"} 1
+				`, provider1, provider2),
+		},
+	}
+	metrics.DekSourceCacheSize.Reset()
+
+	for _, tt := range testCases {
+		t.Run(tt.desc, func(t *testing.T) {
+			metrics.DekSourceCacheSize.Reset()
+			transformer1 := NewEnvelopeTransformer(envelopeService, provider1,
+				func() (State, error) { return *testState.Load(), nil },
+			)
+			kmsv2Transformer1 := value.PrefixTransformer{Prefix: []byte("k8s:enc:kms:v2:"), Transformer: transformer1}
+
+			transformer2 := NewEnvelopeTransformer(envelopeService, provider2,
+				func() (State, error) { return state2, nil },
+			)
+			kmsv2Transformer2 := value.PrefixTransformer{Prefix: []byte("k8s:enc:kms:v2:"), Transformer: transformer2}
+
+			prefix1 := value.NewPrefixTransformers(nil, kmsv2Transformer1)
+			prefix2 := value.NewPrefixTransformers(nil, kmsv2Transformer2)
+
+			envelopeService.keyVersion = tt.keyVersionFromEncrypt
+			var wg sync.WaitGroup
+			for i := 0; i < 10; i++ {
+				i := i // to avoid data race
+				wg.Add(4)
+				for j := 0; j < 3; j++ {
+					j := j // to avoid data race
+					go func() {
+						defer wg.Done()
+						// using different states to ensure we get more than one cache key for a given transformer
+						if j%3 == 0 {
+							testState.Store(&state1)
+						} else if j%3 == 1 {
+							testState.Store(&state2)
+						} else {
+							testState.Store(&state3)
+						}
+						transformedData, err := prefix1.TransformToStorage(ctx, []byte(fmt.Sprintf("%s%d", testText, i)), dataCtx)
+						if err != nil {
+							t.Fatal(err)
```

Comment:
In tests with go routines you cannot use `t` so just `panic`:

```suggestion
							panic(err)
```